### PR TITLE
C++ Add transform support

### PR DESCRIPTION
**Pull request by rleigh-dundee**
- Add transform() functions to OMETransform.h; these use the ome::common::xsl Xerces wrappers
- Add ome-xml transform testcase to test upgrades and downgrades to the current schema version
- Add metadatatools createOMEXMLMetadata testcases to test automatic upgrade and downgrade with the same samples used by the transform tests
- Update OMETIFFReader to cope with badly transformed OME-XML which has lost the file UUID
- Update functions in MetadataTools to correctly apply transforms when required
- Update documentation now transforms are supported

- Remove broken sample file (non-validating; others still remaining are valid according to the schema-but internally inconsistent)
- Correct historical UUID/Creator attribute stripping on upgrades and downgrades; this will make OME-TIFF files readable on upgrade or downgrade without special hacks, and will make multi-file OME-TIFF files readable on upgrade or downgrade (this was completely broken)

- Upgrade ant and maven xerces version to 2.7.2.

Depends on https://github.com/ome/ome-common-cpp/pull/8

--------

Testing: The `transform` and `metadatatools` tests cover all the upgrades and downgrades from all model versions to the 2013-06 model; check the C++ jobs are green, and also check that all the Java ant and maven builds are green.

It should now be possible to run `./bf-test info image.ome.tiff`, where `image.ome.tiff` is using any schema version; it should be upgraded or downgraded appropriately and display without problems, so long as the original file was valid and uncorrupted.

The fixes to the transforms should be manually verified; I'll write a script to help with this and add it here.  Check that for each transform:
- UUID and Creator are preserved (where applicable; UUID was added in the 2008-02 schema, Creator in the 2011-06 schema and downgrades past these points will strip the unsupported attributes)
- Check that the transformed file validates

### CODE REVIEW

**Comment by rleigh-dundee** [comment]
Try to trigger travis.

**Comment by sbesson** [comment]
:+1: for getting Travis green and also fixing the `UUID` as part of this work. Regarding 8643803, is this required for the transforms to work or is it part of the general dependency review/upgrade? If included as part of the PR, we might want to update the Maven versions in the various `pom.xml` as well.

**Comment by rleigh-dundee** [comment]
The xalan upgrade is not required (I think); I did it to fix the original problem but left it in since it fixes a security issue.  It can be split out into a separate PR if needed.  Or I can just fix up maven in this PR.

The DEV jobs are all red.  Working locally but missing something related to xslt include paths.  I'll fix this on or before Monday.

**Comment by rleigh-dundee** [comment]
Green in https://ci.openmicroscopy.org/view/DEV/job/BIOFORMATS-CPP-DEV-merge-superbuild/ and https://ci.openmicroscopy.org/view/DEV/job/BIOFORMATS-CPP-DEV-merge-win-superbuild/ and https://ci.openmicroscopy.org/view/DEV/job/BIOFORMATS-CPP-DEV-merge/ (minus perch)

**Comment by sbesson** [comment]
Tested using the Bio-Formats commmand line tools, the transforms do not seem to preserve the `UUID` :

```
sbesson@lifesci-42275:Downloads $ bfconvert -version
Version: 5.1.8
VCS revision: 9eb8da31c1f01989f439963b9cc215b4398bedd8
Build date: 15 February 2016
sbesson@lifesci-42275:Downloads $ bfconvert test.fake test.ome.tiff
test.fake
FakeReader initializing test.fake
[Simulated data] -> test.ome.tiff [OME-TIFF]
	Converted 1/1 planes (100%)
[done]
0.667s elapsed (12.0+284.0ms per plane, 295ms overhead)
sbesson@lifesci-42275:Downloads $ tiffcomment test.ome.tiff | xmlindent 
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Warning: this comment is an OME-XML metadata block, which contains crucial dimensional parameters and other important metadata. Please edit cautiously (if at all), and back up the original data before doing so. For more information, see the OME-TIFF web site: http://www.openmicroscopy.org/site/support/ome-model/ome-tiff/. -->
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2015-01" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" UUID="urn:uuid:9439287b-2eb0-4586-ab74-accffb153b52" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2015-01 http://www.openmicroscopy.org/Schemas/OME/2015-01/ome.xsd">
   <Image ID="Image:0" Name="test">
      <Pixels BigEndian="false" DimensionOrder="XYZCT" ID="Pixels:0" Interleaved="false" SignificantBits="8" SizeC="1" SizeT="1" SizeX="512" SizeY="512" SizeZ="1" Type="uint8">
         <Channel ID="Channel:0:0" SamplesPerPixel="1">
            <LightPath/>
         </Channel>
         <TiffData FirstC="0" FirstT="0" FirstZ="0" IFD="0" PlaneCount="1">
            <UUID FileName="test.ome.tiff">
               urn:uuid:9439287b-2eb0-4586-ab74-accffb153b52
            </UUID>
         </TiffData>
      </Pixels>
   </Image>
</OME>

sbesson@lifesci-42275:Downloads $ ./bftools/bfconvert -version
Version: 5.2.0-DEV
VCS revision: 3709d5cb1f48607beaf744aa717f85f68fc302dc
Build date: 22 February 2016
sbesson@lifesci-42275:Downloads $ ./bftools/bfconvert test.ome.tiff test2.ome.tiff
test.ome.tiff
OMETiffReader initializing test.ome.tiff
Reading IFDs
Populating metadata
[OME-TIFF] -> test2.ome.tiff [OME-TIFF]
Reading IFDs
Populating metadata
	Converted 1/1 planes (100%)
[done]
0.834s elapsed (22.0+254.0ms per plane, 538ms overhead)
sbesson@lifesci-42275:Downloads $ tiffcomment test2.ome.tiff | xmlindent
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Warning: this comment is an OME-XML metadata block, which contains crucial dimensional parameters and other important metadata. Please edit cautiously (if at all), and back up the original data before doing so. For more information, see the OME-TIFF web site: http://www.openmicroscopy.org/site/support/ome-model/ome-tiff/. -->
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-DEV0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Creator="OME Bio-Formats 5.2.0-DEV" UUID="urn:uuid:f87c6987-e801-4d67-b820-b4ae328108e3" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-DEV0 http://www.openmicroscopy.org/Schemas/OME/2016-DEV0/ome.xsd">
   <Image ID="Image:0" Name="test">
      <Pixels BigEndian="false" DimensionOrder="XYZCT" ID="Pixels:0" Interleaved="false" SignificantBits="8" SizeC="1" SizeT="1" SizeX="512" SizeY="512" SizeZ="1" Type="uint8">
         <Channel ID="Channel:0:0" SamplesPerPixel="1">
            <LightPath/>
         </Channel>
         <TiffData FirstC="0" FirstT="0" FirstZ="0" IFD="0" PlaneCount="1">
            <UUID FileName="test2.ome.tiff">
               urn:uuid:f87c6987-e801-4d67-b820-b4ae328108e3
            </UUID>
         </TiffData>
      </Pixels>
   </Image>
</OME>
```

As this feature was previously broken, this is not a blocker to get this PR merged but might be investigated soon in order to fix our OME-TIFF samples upgrade process. /cc @mtbc

File: cpp/test/ome-bioformats/metadatatools.cpp [hunk 1]
```diff
@@ -553,6 +558,128 @@ TEST_P(CorrectionTest, ValidateAndCorrectModel)
...
+                      continue;
+                    // 2010-06/one-of-everything.xml - parse error
+                    if (schemadir.filename() == path("2010-06") &&
+                        p.file.filename() == path("one-of-everything.xml"))
+                      continue;
+                    // 2010-06/one-of-everything-no-problems.xml - parse error
+                    if (schemadir.filename() == path("2010-06") &&
+                        p.file.filename() == path("one-of-everything-no-problems.xml"))
+                      continue;
+                    // 2013-06/2013-06-datetests.ome - Contains non-POSIX timestamps.
```
Comment (sbesson) [inline cpp/test/ome-bioformats/metadatatools.cpp]: As per the comment in the sample files themselves, I understand these files are valid with regard to the schema. Is this exclusion related to specific requirements in C++? NB: in all cases, we are talking about edge cases.

Comment (rleigh-dundee) [inline cpp/test/ome-bioformats/metadatatools.cpp]: They are valid according to the schema, but internally inconsistent for the most part.  For example, there's a discrepancy between sizeC and samplesPerPixel for each channel.  The actual failing function is this:

https://github.com/openmicroscopy/bioformats/blob/develop/cpp/lib/ome/bioformats/MetadataTools.cpp#L358

These are consistency checks above and beyond schema validation (it can also optionally correct them).  This is based on corrections applied in the Java code, though it might not be in a single function as here.  Ideally, the samples would be sufficiently correct to pass this validation as well.

**Comment by rleigh-dundee** [comment]
@sbesson I think you might be testing the wrong thing with regard to the UUID in your test above.  This is only about testing on reading when we do the schema upgrade.  In your case, you're additionally writing, and the writer is at liberty to rewrite all the UUIDs.  I'd suggest testing the transform step separately by using Xalan or xsltproc on the command-line, and verifying that the new file has a UUID/Creator on the root node.

Example: `Xalan 2013-sample.xml components/specification/transforms/2013-06-to-2015-01.xsl > 2015-sample.xml`

This will test the specific change made here, rather than complicating it with all the additional behaviour of the readers, writers and model/metadata mangling which takes place.

**Comment by sbesson** [comment]
Understood, I misunderstood the scope of the UUID on upgrade fix. I will have another go with the command you pasted above

**Comment by sbesson** [comment]
While testing the UUID, I realized some of the companion datasets are broken. Added to https://trello.com/c/gskA0JKM/65-schema-samples-generation which we can use as a placeholder for fixing our sample files.

Without this PR 
```
 $ xsltproc /opt/ome/bioformats/components/specification/transforms/2015-01-to-2013-06.xsl  companion/multifile-Z1.ome.xml 
<?xml version="1.0"?>
<!-- Warning: this comment is an OME-XML metadata block, which contains crucial dimensional parameters and other important metadata. Please edit cautiously (if at all), and back up the original data before doing so. For more information, see the OME-TIFF web site: http://ome-xml.org/wiki/OmeTiff. -->
<OME:OME xmlns:OME="http://www.openmicroscopy.org/Schemas/OME/2013-06" xmlns:Bin="http://www.openmicroscopy.org/Schemas/BinaryFile/2013-06" xmlns:SPW="http://www.openmicroscopy.org/Schemas/SPW/2013-06" xmlns:SA="http://www.openmicroscopy.org/Schemas/SA/2013-06" xmlns:ROI="http://www.openmicroscopy.org/Schemas/ROI/2013-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2013-06    http://www.openmicroscopy.org/Schemas/OME/2013-06/ome.xsd">
  <BinaryOnly xmlns="http://www.openmicroscopy.org/Schemas/OME/2013-06" MetadataFile="multifile-companion.ome" UUID="urn:uuid:07504f88-7bc3-11e0-b937-2faf67bc00b3"/>
</OME:OME>
```

With this PR:

```
sbesson@ls30630:data_repo $ xsltproc /opt/ome/bioformats/components/specification/transforms/2015-01-to-2013-06.xsl  companion/multifile-Z1.ome.xml 
<?xml version="1.0"?>
<!-- Warning: this comment is an OME-XML metadata block, which contains crucial dimensional parameters and other important metadata. Please edit cautiously (if at all), and back up the original data before doing so. For more information, see the OME-TIFF web site: http://ome-xml.org/wiki/OmeTiff. -->
<OME:OME xmlns:OME="http://www.openmicroscopy.org/Schemas/OME/2013-06" xmlns:Bin="http://www.openmicroscopy.org/Schemas/BinaryFile/2013-06" xmlns:SPW="http://www.openmicroscopy.org/Schemas/SPW/2013-06" xmlns:SA="http://www.openmicroscopy.org/Schemas/SA/2013-06" xmlns:ROI="http://www.openmicroscopy.org/Schemas/ROI/2013-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2013-06    http://www.openmicroscopy.org/Schemas/OME/2013-06/ome.xsd" UUID="urn:uuid:4978087c-a670-4b12-af53-256c62d8d101">
  <BinaryOnly xmlns="http://www.openmicroscopy.org/Schemas/OME/2013-06" MetadataFile="multifile-companion.ome" UUID="urn:uuid:07504f88-7bc3-11e0-b937-2faf67bc00b3"/>
</OME:OME>
```

Additionally the transformation tests are green across platforms with the current set of samples. Outside the scope of this PR, how easy would it be to convert the `find_*_tests` logic to recognize an external samples directory like our public OME-TIFF samples?

**Comment by rleigh-dundee** [comment]
Regarding the test logic, it can certainly be changed to optionally (or additionally) use external samples.  Many tests already add the tests dynamically from a certain location e.g. the samples directory, so it's easy to extend that.

**Comment by dgault** [comment]
New cpp tests and integration of transform support also look good. Using a number of sample files locally the file upgrade and downgrade tests are all running successfully and transform changes appear to be correct for UUID and Creator. File validation also passes as expected. If the full repo is also green then I am happy with these changes.
